### PR TITLE
Chunk + Throttle Netty Writes

### PR DESCRIPTION
**Pull request by original-brownbear**
Fixes issues with allocating huge amounts of direct memory in Netty under heavy write load.
The initial issue reported here was that writing a large message to multiple nodes was incurring a large amount of direct memory allocations when writing a large cluster state to multiple nodes, but there were also reports of ES generally consuming a lot of direct memory in heavy bulk request processing in the past sporadically.

The general reason for these problems lies in the fact that we were invoking Netty's `write` method faster than the network/other side could read and with large individual messages.
* Large individual messages are a problem, as they force Netty to copy the whole message to direct memory before writing if the message is backed by a `byte[]` initially. So situations like writing 100MB to 50 connections simultaneously could result in allocating up to `5G` if writes and/or GCing of buffers are slow.
* Writing faster than the network can transport the data is a problem for the same reason as large writes are a problem: Netty will copy every message received by its `write` method to direct memory and queue it to be written without ever blocking the caller to `write`, which can lead to runaway allocations if writes are initiated faster than they can physically be written long term.

The fix here addresses both issues:

* Only write if the connection signals that it is writable (defined as being below a certain amount of queues writes and the underlying channel being writable in the `OP_WRITE` sense).
* Manually copy to small direct byte buffers right before invoking `write` to avoid big allocations and only allocate buffers for what can be immediately written.
   * Also, small buffers pool much more efficiently than larger ones which are often not pooled at all.

--------------------

We could potentially do even better here memory wise (if necessary) by pooling the direct `ByteBuf` used for writing ourselves instead of using the Netty allocator, but we'd need data from benchmarking to support and measure the impact of that kind of effort.

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2
Jenkins run elasticsearch-ci/packaging-sample

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 1]
```diff
@@ -68,4 +77,111 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
+                currentWrite.promise.tryFailure(new ClosedChannelException());
+            }
+            WriteOperation queuedWrite;
+            while ((queuedWrite = queuedWrites.poll()) != null) {
+                queuedWrite.promise.tryFailure(new ClosedChannelException());
+            }
+            return;
+        }
+        while (channel.isWritable()) {
+            if (currentWrite == null) {
```
Comment (henningandersen) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: Are we sure that there is only one thread in here at a time? If not, we need to guard the reads/writes to currentWrite.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 2]
```diff
@@ -68,4 +77,111 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
+        doFlush(ctx);
+        super.channelInactive(ctx);
+    }
+
+    private void doFlush(ChannelHandlerContext ctx) {
+        final Channel channel = ctx.channel();
+        if (channel.isActive() == false) {
+            if (currentWrite != null) {
+                currentWrite.promise.tryFailure(new ClosedChannelException());
+            }
```
Comment (henningandersen) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: Maybe set currentWrite to null to only trigger once?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 3]
```diff
@@ -68,4 +77,111 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
+            final ByteBuf writeBuffer;
+            if (write.buf.readableBytes() == 0) {
+                write.promise.trySuccess();
+                currentWrite = null;
+                continue;
+            }
+            final int bufferSize = Math.min(write.buf.readableBytes(), 1 << 18);
+            writeBuffer = ctx.alloc().directBuffer(bufferSize, bufferSize);
+            writeBuffer.writeBytes(write.buf, bufferSize);
+            writeBuffer.retain();
```
Comment (henningandersen) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: I think this retain means ref count becomes 2 and thus buffer is not released to pool after write?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 4]
```diff
@@ -68,4 +77,111 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
+                        write.promise.tryFailure(f.cause());
+                    } else {
+                        doFlush(ctx);
+                    }
+                });
+            }
+            ctx.flush();
+            if (channel.isActive() == false) {
+                WriteOperation queuedWrite;
+                while ((queuedWrite = queuedWrites.poll()) != null) {
```
Comment (henningandersen) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: I think we should also notify currentWrite if it is not null and thus this is the same as is done in the beginning of this method (extract to method?).

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 5]
```diff
@@ -68,4 +77,111 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
             transport.onException(tcpChannel, (Exception) newCause);
         }
     }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        assert msg instanceof ByteBuf;
+        final boolean queued = queuedWrites.offer(new WriteOperation((ByteBuf) msg, promise));
```
Comment (henningandersen) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: Are we sure of the lifetime of the msg here? Can it be modified or deallocated once this method returns?

**Review by henningandersen — COMMENTED** [review]
Thanks for your work on this @original-brownbear . I have left comments - please bear in mind that I may be wrong on all of them since this is new territory for me.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java (hunk 5 above)
Comment (henningandersen) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: Sorry, this is OK.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java (hunk 3 above)
Comment (henningandersen) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: I think this is OK, but is the retain and release at all necessary? Otherwise the retain should move to after ctx.write to null out the release when write is successful?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 6]
```diff
@@ -68,4 +77,111 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+        Channel channel = ctx.channel();
+        if (channel.isWritable() || channel.isActive() == false) {
+            doFlush(ctx);
+        }
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: `ChannelInboundHandler` has state callbacks for the channel in general, it's not specific to reading if that's what makes you suspicious? :)

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java (hunk 1 above)
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: We shouldn't be calling `flush` from outside the event loop ever an all other invocations of this method do happen in event loop callbacks so we should be good here. If you look at other Netty handler implementations (like compression implementations or so) you'll see that they make the same assumption about flush.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java (hunk 3 above)
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: One reference is released by  `ctx.write(writeBuffer)` :)

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 7]
```diff
@@ -68,4 +77,111 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
+            }
+            final WriteOperation write = currentWrite;
+            final ByteBuf writeBuffer;
+            if (write.buf.readableBytes() == 0) {
+                write.promise.trySuccess();
+                currentWrite = null;
+                continue;
+            }
+            final int bufferSize = Math.min(write.buf.readableBytes(), 1 << 18);
+            writeBuffer = ctx.alloc().directBuffer(bufferSize, bufferSize);
```
Comment (henningandersen) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: Could we skip the copy if the write is smaller than 1<<18?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java (hunk 4 above)
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: `currentWrite` will have its callback fail on the event loop because we just enqueued a flush for it. If we touch it here as well it should be a noop.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java (hunk 2 above)
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: That's ok `currentWrite.promise.tryFailure` is idempotent and I think it's easier to let Netty guard us here than do it ourselves and maybe make a mistake :)

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java (hunk 7 above)
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: Maybe, it's probably a noop since Netty will itself do the same copying internally if we don't do it here. I mainly did it here to make profiling a little easier (only one source of allocations for writes). 

Also, I think we could do even better here performance wise if we want/need to by using a different allocator that just caches a single direct `ByteBuf`, so I figured this way of extracting the alloc to a single place makes it a little easier to experiment with that kind of thing. I'll add some more comments on that to the PR description in a bit.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java (hunk 3 above)
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: @henningandersen cleaned up the redundant retain release dance here, thanks for noticing :)

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 8]
```diff
@@ -37,12 +42,18 @@
 
     private final Netty4Transport transport;
 
+    private final ConcurrentLinkedQueue<WriteOperation> queuedWrites = new ConcurrentLinkedQueue<>();
+
+    private volatile ChannelHandlerContext ctx;
+
+    private WriteOperation currentWrite;
+
     Netty4MessageChannelHandler(Netty4Transport transport) {
```
Comment (tbrooks8) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: I'm not sure that this should be in this (`Netty4MessageChannelHandler`). You are copying heap bytes -> direct bytes. That works fine for plaintext. But this handler comes before the `SslHandler`. Which means in the security use case, this is heap bytes -> direct bytes -> heap bytes (temporary byte array in `SSLEngine`) encrypted->  heap buffer -> direct bytes (in `filterOutboundMessages`). So this will introduce more copies.

I think that either:

1. This should be a outbound handler that comes after the `SslHandler`
2. This should not copy the bytes to a direct buffer and instead just limit the heap bytes that are passed to the next handler until the writability callback is invoked.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 9]
```diff
@@ -68,4 +79,121 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
+                    } else {
+                        resumeFlush();
+                    }
+                });
+            }
+            ctx.flush();
+            if (channel.isActive() == false) {
+                WriteOperation queuedWrite;
+                while ((queuedWrite = queuedWrites.poll()) != null) {
+                    queuedWrite.promise.tryFailure(new ClosedChannelException());
```
Comment (tbrooks8) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: It is confusing that these are cleared in the `doFlush` call. Can we not just clear and release everything in the `channelInactive` callback? Or the `close` callback if you change this to an outbound handler and put it after the ssl handler?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 10]
```diff
@@ -68,4 +79,121 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
+            return;
+        }
+        while (channel.isWritable()) {
+            if (currentWrite == null) {
+                currentWrite = queuedWrites.poll();
+            }
+            if (currentWrite == null) {
+                break;
+            }
+            final WriteOperation write = currentWrite;
```
Comment (tbrooks8) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: I think you will need to guard against the promise from already failing. The prior channel write could have failed and failed the the promise associated with the write operation. I imagine that this will kill the channel and cause everything to shutdown. But it still seems safest to prevent potentially flushing bytes that have already been released.

**Review by Tim-Brooks — CHANGES_REQUESTED** [review]
Some comments.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 11]
```diff
@@ -68,4 +79,121 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
+            final int bufferSize = Math.min(write.buf.readableBytes(), 1 << 18);
+            final ByteBuf writeBuffer = ctx.alloc().directBuffer(bufferSize, bufferSize);
+            writeBuffer.writeBytes(write.buf, bufferSize);
+            final ChannelFuture writeFuture = ctx.write(writeBuffer);
+            if (write.buf.readableBytes() == 0) {
+                currentWrite = null;
+                writeFuture.addListener((ChannelFutureListener) future -> {
+                    if (future.isSuccess()) {
+                        write.promise.trySuccess();
+                        resumeFlush();
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: Right, this was coming from the mistaken assumption that the future may resolve in a different thread -> removed :)

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 12]
```diff
@@ -68,4 +79,121 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
+        doFlush(ctx);
+        super.channelInactive(ctx);
+    }
+
+    private void resumeFlush() {
+        final ChannelHandlerContext context = this.ctx;
+        if (context == null) {
+            return;
+        }
+        if (context.executor().inEventLoop()) {
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: Right :) my mistake, sorry for the noise :)

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java (hunk 9 above)
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: This solves the problem of potentially trying to flush released bytes because we can't just rely on the channel inactive hook as far as I can see because it triggers in a later iteration of the event-loop. From `io.netty.channel.AbstractChannel.AbstractUnsafe#close`:

```java
            // Fail all the queued messages
            try {
                outboundBuffer.failFlushed(CLOSED_CHANNEL_EXCEPTION);
                outboundBuffer.close(CLOSED_CHANNEL_EXCEPTION);
            } finally {

                if (wasActive && !isActive()) {
                    invokeLater(new OneTimeTask() {
                        @Override
                        public void run() {
                            pipeline.fireChannelInactive();
                        }
                    });
                }
```

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java (hunk 10 above)
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: That's why I put the cleanup loop that you mentioned [above](https://github.com/elastic/elasticsearch/pull/39286#discussion_r261421868). This way we def. fail all the writes and drain the queue on failures before going for another write in the loop. And since the only way, the callbacks/futures that would release the bytes is the `ctx.flush` we do at the end of the loop body I think we're safe here?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java (hunk 8 above)
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java]: Right :) I went with option 2. now. That also makes the whole issue of [potentially trying to write released bytes](https://github.com/elastic/elasticsearch/pull/39286#discussion_r261422662) a lot safer/cleaner. Once/if the buffer is below our write chunk size we can simply pass it down outright and use retained slices in the writes preceding that situation.
(+ we save some objects for writes that are smaller than the buffer size :))

**Comment by original-brownbear** [comment]
@tbrooks8 thanks for pointing out the threading situation :) (my bad for getting confused by the SslHandler here :(), I simplified the logic accordingly.

**Comment by travishaagen** [comment]
@original-brownbear , instead of creating new instances of `ClosedChannelException`,
I suggest you consider instantiating a `static final` instance that can be
reused. I was first introduced to the idea here:

http://normanmaurer.me/blog/2013/11/09/The-hidden-performance-costs-of-instantiating-Throwables/

You can see from the search results below that the Netty project often uses this technique,
but I did see that they sometimes do not:

https://github.com/netty/netty/search?q=new+ClosedChannelException%28%29&unscoped_q=new+ClosedChannelException%28%29

In a nutshell, this will reduce object creation and eliminate the performance cost of
instantiating a `Throwable`, where the cost is from the stacktrace generation. As a 
rule of thumb, whenever a `Throwable` is used for flow-control it may be a good idea. 
If you care about the stacktrace for logging, then you would not want to do this.

**Comment by original-brownbear** [comment]
@travishaagen since the connections here are intended to be long-lived and the messages generally have a number of other objects attached for their callbacks I don't think we have to worry about the cost of setting up these `ClosedChannelExceptions` here.

**Comment by original-brownbear** [comment]
@tbrooks8 ping :) can you take another look, please? Thanks!

**Comment by tbrooks8** [comment]
Yes I will take a look tomorrow.

**Review by Tim-Brooks — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
@tbrooks8 thanks!
